indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexSubTask.java

```diff
@@ -195,63 +205,155 @@ public TaskStatus run(final TaskToolbox toolbox) throws Exception
+  }
+
+  @Override
+  public List<DataSegment> findInputSegments(TaskActionClient taskActionClient, List<Interval> intervals)
       throws IOException
   {
-    for (Interval interval : Tasks.computeCompactIntervals(intervals)) {
-      Preconditions.checkNotNull(
-          client.submit(
-              new SurrogateAction<>(supervisorTaskId, new LockTryAcquireAction(TaskLockType.EXCLUSIVE, interval))
-          ),
-          "Cannot acquire a lock for interval[%s]", interval
-      );
-    }
+    return IndexTask.findInputSegments(
+        getDataSource(),
+        taskActionClient,
+        intervals,
+        ingestionSchema.getIOConfig().getFirehoseFactory()
+    );
   }
 
-  private SegmentAllocator createSegmentAllocator(
-      TaskToolbox toolbox,
-      ParallelIndexTaskClient taskClient,
-      ParallelIndexIngestionSpec ingestionSchema
-  )
+  @Override
+  public boolean checkIfChangeSegmentGranularity(List<Interval> intervalOfExistingSegments)
   {
-    final DataSchema dataSchema = ingestionSchema.getDataSchema();
-    final ParallelIndexIOConfig ioConfig = ingestionSchema.getIOConfig();
-    if (ioConfig.isAppendToExisting()) {
-      return new ActionBasedSegmentAllocator(
-          toolbox.getTaskActionClient(),
-          dataSchema,
-          (schema, row, sequenceName, previousSegmentId, skipSegmentLineageCheck) -> new SurrogateAction<>(
-              supervisorTaskId,
-              new SegmentAllocateAction(
-                  schema.getDataSource(),
-                  row.getTimestamp(),
-                  schema.getGranularitySpec().getQueryGranularity(),
-                  schema.getGranularitySpec().getSegmentGranularity(),
-                  sequenceName,
-                  previousSegmentId,
-                  skipSegmentLineageCheck
-              )
-          )
-      );
+    final Granularity segmentGranularity = ingestionSchema.getDataSchema().getGranularitySpec().getSegmentGranularity();
+    return intervalOfExistingSegments.stream().anyMatch(interval -> !segmentGranularity.match(interval));
+  }
+
+  @Override
+  public boolean isPerfectRollup()
+  {
+    return false;
+  }
+
+  @Nullable
+  @Override
+  public Granularity getSegmentGranularity(Interval interval)
+  {
+    final GranularitySpec granularitySpec = ingestionSchema.getDataSchema().getGranularitySpec();
+    if (granularitySpec instanceof ArbitraryGranularitySpec) {
+      return null;
     } else {
-      return (row, sequenceName, previousSegmentId, skipSegmentLineageCheck) -> taskClient.allocateSegment(
-          supervisorTaskId,
-          row.getTimestamp()
-      );
+      return granularitySpec.getSegmentGranularity();
+    }
+  }
+
+  @VisibleForTesting
+  SegmentAllocator createSegmentAllocator(TaskToolbox toolbox, ParallelIndexTaskClient taskClient)
+  {
+    return new WrappingSegmentAllocator(toolbox, taskClient);
+  }
+
+  private class WrappingSegmentAllocator implements SegmentAllocator
```

Comment:
Suggest calling this `LazyInitializationSegmentAllocator` instead

Reply:
Hmm, I'm not sure the suggested name is better. It's unclear to me what is initialized lazily.